source-firestore/pull.go

```diff
@@ -0,0 +1,322 @@
+package main
+
+import (
+	firestore "cloud.google.com/go/firestore"
+	"context"
+	"encoding/json"
+	"errors"
+	firebase "firebase.google.com/go"
+	"fmt"
+	pc "github.com/estuary/flow/go/protocols/capture"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	"io"
+	"time"
+)
+
+type capture struct {
+	Stream   pc.Driver_PullServer
+	Bindings []resource
+	Config   config
+	Acks     <-chan *pc.Acknowledge
+
+	// The channel where documents will be sent to from different streams
+	// this is here to ensure we write documents sequentially and avoid scrambled outputs
+	Output chan emitRequest
+
+	Tail bool
+}
+
+type EmitterRequestType int8
+
+const (
+	EncodeDocument EmitterRequestType = 0
+	StateUpdate    EmitterRequestType = 1
+)
+
+// struct sent over channel to emitter
+type emitRequest struct {
+	requestType EmitterRequestType
+	arena       []byte
+	binding     uint32
+	state       []byte
+}
+
+func (driver) Pull(stream pc.Driver_PullServer) error {
```

Comment:
no action: a little more boilerplate would go a long way to simplify this further...

Reply:
Will already acknowledged that we can implement something similar to RunTransactions here as well

Reply:
I still intend to implement some better boilerplate around the Pull RPC stuff, but I'd prefer to get this PR merged with as few additional changes as possible and then figure out some better boilerplate later.